File: internal/api/dto/task.go
package dto

import (
	"time"

	"github.com/ZeusWPI/events/internal/db/model"
	"github.com/ZeusWPI/events/internal/pkg/task"
)

// TaskHistoryStatus are the possible results a TaskHistory
type TaskHistoryStatus string

const (
	// Success indicates the task succeeded
	Success TaskHistoryStatus = "Success"
	// Failed indicates the task returned an error
	Failed TaskHistoryStatus = "Failed"
)

// TaskHistory is the data transferable object version of the model Task
type TaskHistory struct {
	ID        int               `json:"id"`
	Name      string            `json:"name"`
	Result    TaskHistoryStatus `json:"result"`
	RunAt     time.Time         `json:"run_at"`
	Error     string            `json:"error,omitempty"`
	Recurring bool              `json:"recurring"`
}

// TaskHistoryDTO converts a model Task to a DTO TaskHistory
func TaskHistoryDTO(task *model.Task) TaskHistory {
	taskError := ""
	if task.Error != nil {
		taskError = task.Error.Error()
	}

	return TaskHistory{
		ID:        task.ID,
		Name:      task.Name,
		Result:    TaskHistoryStatus(task.Result),
		RunAt:     task.RunAt,
		Error:     taskError,
		Recurring: task.Recurring,
	}
}

// TaskHistoryFilter is the filter used for task histories
type TaskHistoryFilter struct {
	Name        string
	OnlyErrored bool
	Recurring   *bool
	Page        int
	Limit       int
}

// TaskStatus are the different statuses a Task can have
type TaskStatus string

const (
	// Running indicates the task is currently running
	Running TaskStatus = "Running"
	// Waiting indicates the task is currently waiting to be executed
	Waiting TaskStatus = "Waiting"
)

// Task is the data transferable object version of a task.Stat returned by the task manager
type Task struct {
	ID         int               `json:"id"`
	Name       string            `json:"name"`
	Status     TaskStatus        `json:"status"`
	NextRun    time.Time         `json:"next_run"`
	Recurring  bool              `json:"recurring"`
	LastStatus TaskHistoryStatus `json:"last_status,omitempty"`
	LastRun    *time.Time        `json:"last_run,omitempty"`
	LastError  string            `json:"last_error,omitempty"`
	Interval   *time.Duration    `json:"interval,omitempty"`
}

// TaskDTO converts a task.Stat to a DTO Task
func TaskDTO(task task.Stat) Task {
	t := Task{
		ID:        task.ID,
		Name:      task.Name,
		Status:    TaskStatus(task.Status),
		NextRun:   task.NextRun,
		Recurring: task.Recurring,
	}

	if task.Recurring {
		t.LastStatus = TaskHistoryStatus(task.LastStatus)
		t.LastRun = &task.LastRun
		t.LastError = ""
		if task.LastError != nil {
			t.LastError = task.LastError.Error()
		}
		t.Interval = &task.Interval
	}

	return t
}

File: ui/src/lib/types/task.ts
import type { API } from "./api";
import { Base } from "./general";

export enum TaskStatus {
  Waiting = "waiting",
  Running = "running",
}

export enum TaskResult {
  Succes = "success",
  Failed = "failed",
  Resolved = "resolved",
}

export enum TaskType {
  Recurring = "recurring",
  Once = "once",
}

export interface Task {
  uid: string;
  name: string;
  status: TaskStatus;
  nextRun: Date;
  type: TaskType;
  lastStatus?: TaskResult;
  lastRun?: Date;
  lastError?: string;
  interval?: number;
}

export interface TaskHistory extends Base {
  name: string;
  result: TaskResult;
  runAt: Date;
  error?: string;
  type: TaskType;
  duration: number;
}

export interface TaskHistoryFilter {
  uid?: string;
  result?: TaskResult;
}

export function convertTaskToModel(task: API.Task): Task {
  return {
    uid: task.uid,
    name: task.name,
    status: task.status as TaskStatus,
    nextRun: new Date(task.next_run),
    type: task.type as TaskType,
    lastStatus: task.last_status ? task.last_status as TaskResult : undefined,
    lastRun: task.last_run ? new Date(task.last_run) : undefined,
    lastError: task.last_error,
    interval: task.interval,
  };
}

export function convertTasksToModel(tasks: API.Task[]): Task[] {
  return tasks.map(convertTaskToModel);
}

export function convertTaskHistoryToModel(history: API.TaskHistory[]): TaskHistory[] {
  return history.map(history => ({
    id: history.id,
    name: history.name,
    result: history.result as TaskResult,
    runAt: new Date(history.run_at),
    error: history.error,
    type: history.type as TaskType,
    duration: history.duration,
  }));
}
